charms/tensorboard-controller/src/charm.py

```diff
@@ -1,184 +1,302 @@
+        """
+        self.unit.status = MaintenanceStatus("Creating K8s resources")
+        try:
+            self.k8s_resource_handler.apply()
+        except ApiError as error:
+            if self._check_and_report_k8s_conflict(error) and force_conflicts:
+                # conflict detected when applying K8s resources
+                # re-apply K8s resources with forced conflict resolution
+                self.unit.status = MaintenanceStatus("Force applying K8s resources")
+                self.logger.warning("Apply K8s resources with forced changes against conflicts")
+                self.k8s_resource_handler.apply(force=force_conflicts)
+            else:
+                raise GenericCharmRuntimeError("K8s resources creation failed") from error
+        try:
+            self.crd_resource_handler.apply()
+        except ApiError as error:
+            if self._check_and_report_k8s_conflict(error) and force_conflicts:
+                # conflict detected when applying CRD resources
+                # re-apply CRD resources with forced conflict resolution
+                self.unit.status = MaintenanceStatus("Force applying CRD resources")
+                self.logger.warning("Apply CRD resources with forced changes against conflicts")
+                self.crd_resource_handler.apply(force=force_conflicts)
+            else:
+                raise GenericCharmRuntimeError("CRD resources creation failed") from error
+        self.model.unit.status = MaintenanceStatus("K8s resources created")
 
-    def _check_leader(self):
+    def _check_leader(self) -> None:
+        """Check whether a unit is the leader."""
         if not self.unit.is_leader():
             # We can't do anything useful when not the leader, so do nothing.
-            raise CheckFailed("Waiting for leadership", WaitingStatus)
+            self.logger.warning("Not a leader, skipping setup")
+            raise ErrorWithStatus("Waiting for leadership", WaitingStatus)
+
+    def _check_container_connection(self) -> None:
+        """Check if connection can be made with container."""
+        if not self.container.can_connect():
+            raise ErrorWithStatus("Pod startup is not complete", MaintenanceStatus)
+
+    def _check_status(self) -> None:
+        """Check status of workload and set status accordingly."""
+        self._check_leader()
+        container = self.unit.get_container(self._container_name)
+        if container:
+            try:
+                check = container.get_check("tensorboard-controller-up")
+            except ModelError as error:
+                raise GenericCharmRuntimeError(
+                    "Failed to run health check on workload container"
+                ) from error
+            if check.status != CheckStatus.UP:
+                self.logger.error(
+                    f"Container {self._container_name} failed health check. It will be restarted."
```

Comment:
What exactly is restarted. My guess is the workload service? Is it done automatically?
If it is the service, let's reflect that in the messaging.

Reply:
The intention is for the workload service to be restarted, yes! This is possible using the `on-check-failure` service configuration setting (relevant [docs](https://github.com/canonical/pebble#health-checks) section).

I'll make the error message clearer!